packages/interflex/src/__tests__/translate.test.ts

```diff
@@ -0,0 +1,115 @@
+
+describe('translate function', () => {
+  beforeAll(() => {
+    injectDictionaries({
+      en: generateDefaultDictionary({
+        simple: 'Welcome!',
+        interpolate: 'Welcome, {name}.',
+        complex: 'Welcome! Please click <Link>here</Link>.',
+        complexInterpolate: 'Welcome, <foo>{name}</foo>.',
+        nextedComplexInterpolate:
+          'Welcome, <foo>{firstName} <bar>{lastName}</bar></foo>.',
+      }),
+    });
+  });
+
+  it('should be no-op for strings that do not exist in the dictionary', () => {
+    expect(t('')).toBe('');
+    expect(t('Foo')).toBe('Foo');
+    expect(t('Hello, {foo}!')).toBe('Hello, {foo}!');
+    expect(t('welcome', { locale: 'id' })).toBe('welcome');
+  });
+
+  it('should return the correct string that exist in the dictionary', () => {
+    expect(t('simple')).toBe('Welcome!');
+    expect(t('interpolate')).toBe('Welcome, {name}.');
+    expect(t('complex')).toBe('Welcome! Please click <Link>here</Link>.');
+    expect(t('complexInterpolate')).toBe('Welcome, <foo>{name}</foo>.');
+    expect(t('nextedComplexInterpolate')).toBe(
+      'Welcome, <foo>{firstName} <bar>{lastName}</bar></foo>.',
+    );
+  });
+
+  it('should interpolate strings', () => {
+    expect(t('Hello, {foo}!', { params: { foo: 'John' } })).toEqual([
+      'Hello, John!',
+    ]);
+    expect(t('interpolate', { params: { name: 'John' } })).toEqual([
+      'Welcome, John.',
+    ]);
+    expect(t('complexInterpolate', { params: { name: 'John' } })).toEqual([
+      'Welcome, ',
+      '<foo>John</foo>',
+      '.',
+    ]);
+    expect(
+      t('nextedComplexInterpolate', {
+        params: { firstName: 'John', lastName: 'Doe' },
+      }),
+    ).toEqual(['Welcome, ', '<foo>John <bar>Doe</bar></foo>', '.']);
+  });
+
+  it('should interpolate tags', () => {
+    expect(
+      t('complexInterpolate', {
+        params: {
+          foo: (contents: unknown) => ({
+            component: 'Foo',
+            contents,
+          }),
+        },
+      }),
+    ).toEqual(['Welcome, ', { component: 'Foo', contents: ['name'] }, '.']);
+    expect(
+      t('nextedComplexInterpolate', {
+        params: {
+          foo: (contents: unknown) => ({ component: 'Foo', contents }),
+          bar: (contents: unknown) => ({ component: 'Bar', contents }),
+        },
+      }),
+    ).toEqual([
+      'Welcome, ',
+      {
+        component: 'Foo',
+        contents: [
+          'firstName ',
+          { component: 'Bar', contents: ['lastName'] },
+          '',
+        ],
+      },
+      '.',
+    ]);
+  });
+
+  it('should interpolate tags and strings', () => {
+    expect(
+      t('complexInterpolate', {
+        params: {
+          name: 'John',
+          foo: (contents: unknown) => ({
+            component: 'Foo',
+            contents,
+          }),
+        },
+      }),
+    ).toEqual(['Welcome, ', { component: 'Foo', contents: ['John'] }, '.']);
```

Comment:
I wonder how this will be rendered. Do you have any example?

Like when we use it like this:
```tsx
<Text>
  t(
    'My name is <foo>{name}</foo>',
    {
      params: {
        name: 'John',
        foo: (contents: unknown) => ({ component: 'Foo', contents }),
      },
    }
  )
</Text>
```

It will get returned like this, right?
```tsx
<Text>
  {
    [
      'My name is ',
      {  component: 'Foo', contents: ['John'] },
    ]
  }
</Text>
```

Reply:
That is correct. Since the test function returns an object, of-course that would not be rendered meaningfully. In reality, we return a node in the parameter(s).

Something like:
```tsx
<Text>
  t(
    'My name is <foo>{name}</foo>',
    {
      params: {
        name: 'John',
        // NOTE: output will be an array so key is necessary
        foo: (contents) => <Text bold key="foo">{contents}</Text>,
      },
    }
  )
</Text>
```

will return:
```tsx
<Text>
  {
    [
      'My name is ',
      <Text bold key="foo">John</Text>,
    ]
  }
</Text>
```

then it will render as you expect; visually: My name is **John**

For complex, nested tags example:
```tsx
<Text>
  t(
    'My name is <foo>{firstName} <bar>{lastName}</bar></foo>',
    {
      params: {
        firstName: 'John',
        lastName: 'Doe',
        // NOTE: output will be an array so key is necessary
        foo: (contents) => <Text bold key="foo">{contents}</Text>,
        bar: (contents) => <Text italic key="bar">{contents}</Text>,
      },
    }
  )
</Text>
```

will return:
```tsx
<Text>
  {
    [
      'My name is ',
      <Text bold key="foo">
        {
          [
            'John ',
            <Text italic key="bar">Doe</Text>,
          ]
        }
      </Text>,
    ]
  }
</Text>
```

then it will render visually as: My name is **John _Doe_**

Reply:
Tried running the examples above on a `flexship` frontend project:
![image](https://user-images.githubusercontent.com/16997526/79206048-f9bf8c00-7e68-11ea-97c9-a1f6e5266f42.png)

Reply:
Oh, I see. So that function could return anything, not just an array.
Now you mention it, I just realized that `unknown` in `Params` type.

This is really cool! I fell in love with this already!

I think this is really similar to what @sstur suggested.

Reply:
Glad to hear that!

Yes, this is based on @sstur implementation and improved so it can handle nested tags.

Though, the rest of the stuff still needs to be discussed. Like loading and storing dictionaries, attaching translate function to global object and making typescript recognize that, and selecting locale.